### ISSUE TITLE
Use generic functions for Optimizers

### ISSUE THREAD

**Issue by Benny-Nottonson** [Benny-Nottonson/voodoo#20]
**Is your feature request related to a problem? Please describe.**
There are currently only 3 optimizers but they should in theory all be using the same generic structure since they are only modifying the training rate.

**Describe the solution you'd like**
The loss functions should be encapsuled into one generic structure like what was done with Activations.

**Describe alternatives you've considered**
The current implementation is fine but adds unneeded complexity.
